raytracing/drawing.py

```diff
@@ -0,0 +1,324 @@
+
+    @property
+    def hasLabel(self) -> bool:
+        if self.label is None:
+            return False
+        return True
+
+    def applyTo(self, axes: plt.Axes, x: float = None, y: float = None):
+        """ Apply the Drawing on a figure at a given position (x, y) with auto-scale.
+
+        Args:
+            axes (matplotlib.pyplot.Axes): The figure's Axes on which to apply the drawing.
+            x, y (:obj:`float`, optional): The x and y position in data units where to apply the drawing.
+                Defaults to (0, 0).
+
+        """
+
+        self.axes = axes
+        if x is not None:
+            self.x = x
+        if y is not None:
+            self.y = y
+
+        self.update()
+
+        for component in self.components:
+            self.axes.add_patch(component)
+
+        if self.label is not None:
+            self.axes.add_artist(self.label)
+
+    def update(self, x: float = None, y: float = None):
+        """ Update the drawing's position and scaling.
+
+        Args:
+            x, y (:obj:`float`, optional): The x and y position where to apply the drawing.
+                Defaults to the originally applied position.
+
+        """
+
+        if x is not None:
+            self.x = x
+        if y is not None:
+            self.y = y
+
+        xScaling, yScaling = self.scaling()
+
+        translation = transforms.Affine2D().translate(self.x, self.y)
+        scaling = transforms.Affine2D().scale(xScaling, yScaling)
+
+        for component in self.components:
+            component.set_transform(scaling + translation + self.axes.transData)
+
+        if self.label is not None:
+            self.label.set_transform(translation + self.axes.transData)
+
+    def scaling(self):
+        """ Used internally to compute the required scale transform so that the width of the objects stay the same
+        respective to the Axes. """
+        if not self.useAutoScale:
+            return 1, 1
+
+        xScale, yScale = self.axesToDataScale()
+
+        heightFactor = self.halfHeight() * 2 / yScale
+        xScaling = xScale * (heightFactor / 0.2) ** (3 / 4)
+
+        return xScaling, 1
+
+    def axesToDataScale(self):
+        """ Dimensions of the figure in data units. """
+        xScale, yScale = self.axes.viewLim.bounds[2:]
+
+        return xScale, yScale
+
+    def halfHeight(self) -> float:
+        """ Maximum absolute Y-value of the drawing (not affected by the transforms).
+        Used internally to auto-scale. """
+        halfHeight = 0
+        for component in self.components:
+            componentMaxY = np.max(np.abs(component.get_xy()), axis=0)[1]
+            if componentMaxY > halfHeight:
+                halfHeight = componentMaxY
+        return halfHeight
+
+    @property
+    def centroid(self):
+        xy = []
+        for component in self.components:
+            xy.extend(component.get_xy())
+
+        return np.mean(xy, axis=0)
+
+
+class SurfacePairPatch(patches.PathPatch):
```

Comment:
You should then define `SurfacePair`,`Arrow`, `Aperture` and `Label` in general. They can manage most details except the actual drawing. Then you would have two options:
1. specialize them immediately with `SurfacePairPatch`, `ArrowPatch` etc... and draw with matplotlib
or
2. create a `Bezier` class, because that is the most general drawing in 2D: its interface would be general, because a Bezier curve is general. You could then create a `BezierPatch` or an internal variable that would implement the details in matplotlib, or any other backend.

I prefer option 2, it is much less work I think: a new backend would simply mean implementing Bezier.

Reply:
I understand. Option 2 still requires to break down the arrow, doubleArrow, aperture, surfacePair, and future graphics in a list of bezier curves which takes some time.

Reply:
Maybe you misunderstand what I am saying because I don't think it is that much work.
You could also have an option to add arrows at both ends of your Bezier curve.